recipes/opencv/4.x/conanfile.py

```diff
@@ -223,6 +227,7 @@ def _configure_cmake(self):
         self._cmake.definitions["WITH_OPENJPEG"] = self.options.with_jpeg2000 == "openjpeg"
         self._cmake.definitions["WITH_OPENEXR"] = self.options.with_openexr
         self._cmake.definitions["WITH_EIGEN"] = self.options.with_eigen
+        self._cmake.definitions["HAVE_QUIRC"] = self.options.with_quirc  # force usage of quirc reqirement
```

Comment:
seems fine: https://github.com/opencv/opencv/blob/8ac304517e0ec2f9ad695b05b94b1ad1d1e82c1c/CMakeLists.txt#L780

Reply:
Yeah,think so too. In https://docs.opencv.org/4.5.0/db/d05/tutorial_config_reference.html I found this:"Options starting with HAVE_ indicate that dependency have been enabled, can be used to manually enable a dependency if automatic detection can not be used.". Should be safe for future versions.

Reply:
So, this is resolved?

Reply:
Seems too simple, `WITH_QUIRC` is ON by default, it means that vendored quirc lib is still compiled and linked to OpenCV.

Reply:
Ah, no, I just reverted to the original version for this. There is a block where `WITH_` options are all disabled. `WITH_QUIRC` is now there back again :)

Reply:
I believe that `HAVE_QUIRC` is reset by `OCV_OPTION` macro:
https://github.com/opencv/opencv/blob/8ac304517e0ec2f9ad695b05b94b1ad1d1e82c1c/CMakeLists.txt#L434
https://github.com/opencv/opencv/blob/de385009ae9aa84f3c93e1322a541e2dd6308900/cmake/OpenCVUtils.cmake#L727

Otherwise, looking at build files, I don't understand how it could build, because opencv_objdetect is linked to `quirc` target, and this target can't come from a conan generator (and opencv build files never tries to find an external quirc through find_package or pkg-config).

Reply:
~/.conan/data/opencv/4.5.0/_/_/build/64f116605da90b9fd346a277849fb125abbc871e/build_subfolder/CMakeCache.txt:

```
...
HAVE_QUIRC:UNINITIALIZED=True
...
WITH_QUIRC:BOOL=False
...
opencv_objdetect_LIB_DEPENDS:STATIC=...;general;quirc;
...
```

~/.conan/data/opencv/4.5.0/_/_/package/64f116605da90b9fd346a277849fb125abbc871e/lib:

```
libopencv_calib3d.a
libopencv_core.a
libopencv_features2d.a
libopencv_flann.a
libopencv_highgui.a
libopencv_imgcodecs.a
libopencv_imgproc.a
libopencv_ml.a
libopencv_objdetect.a
libopencv_photo.a
libopencv_stitching.a
libopencv_video.a
libopencv_videoio.a
```

As you can see, the variables are set as expected in `CMakeCache.txt`. Furthermore, no more `libquirc.a` is in the package lib dir.


The external dependeny is appended to `opencv_objdetect` because I added the code for this similar to other requirements:

```python
...
        if self.options.with_quirc:
            self.requires("quirc/1.1")
...
        def quirc():
            return ["quirc::quirc"] if self.options.with_quirc else []
...
            {"target": "opencv_objdetect",  "lib": "objdetect",  "requires": [...] + eigen() + quirc()},
...

```
 

I ran it with my current project too, where I'm using cv::QrCodeDetector, and it works. 

Regarding the `OCV_OPTION` macro, I didn't study this in detail. It looks quite complicated to me. But maybe the key is the usage of the `option()` function.

cmake --help-command option

```
option
------

Provide an option that the user can optionally select.

 option(<variable> "<help_text>" [value])

Provides an option for the user to select as ``ON`` or ``OFF``.
If no initial ``<value>`` is provided, ``OFF`` is used.
If ``<variable>`` is already set as a normal or cache variable,
then the command does nothing (see policy ``CMP0077``).

If you have options that depend on the values of other options, see
the module help for ``CMakeDependentOption``.
```

I would say this means, if `HAVE_QUIRC` is set with `-D` command line option, the call to `option()` does not override the users value ("If \<variable\> is already set as a normal or cache variable, **then the command does nothing**").

So my résumé is, that it works as expected. I hope I modified the `conanfile.py` in a clean way?

Reply:
Indeed, it seems to work.
External quirc is linked into `opencv_objdetect`. I don't know how, maybe with the magic of `conan_basic_setup()`?

Reply:
Wow you're really digging deep into conan's and CMake's implementation details in search for understanding. Good to know that everything works by intention and not only by accident, haha. My goal was to extend the existing recipe in a way with minimal impact for the rest of the recipe and and reuse the existing infrastructure.

OK, I added `quirc` to the list of diretories to remove in `3rdparty` in `_patch_opencv()`. I missed this part, but I agree it makes totally sense to remove this as an extra safety net. And to stay consistent with the whole idea of the recipe.